### PR TITLE
3568: Add logout logic

### PR DESCRIPTION
**Pull request by kat1906**

### CODE REVIEW

File: app/config.js [hunk 1]
```diff
@@ -6,10 +6,10 @@ module.exports = {
   env: process.env.NODE_ENV || 'development',
 
   // Port to run local development server on
-  port: process.env.PORT || 3000,
+  port: process.env.PORT || 3001,
```
Comment (alibombaye) [inline app/config.js]: why has this changed?

File: app/pages/homepage/ui.test.js [hunk 2]
```diff
@@ -1,51 +1,60 @@
...
-  if (withAuth) {
-    await setCookies();
-  }
+const setLoggedInCookies = ClientFunction(() => {
+  document.cookie = 'cookie1=cookie1;';
+});
+
+const getCookies = ClientFunction(() => document.cookie);
+
+const pageSetup = async ({ t, withAuth = false, withLoggedInCookies = false }) => {
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: I don't think we need both here. They are both serving the same purpose which is setting a cookie on the document to replicate being logged in

File: app/pages/homepage/ui.test.js [hunk 3]
```diff
@@ -54,47 +63,98 @@ test('should navigate to home page header banner', async (t) => {
...
   await t
     .expect(await extractInnerText(logoutComponent)).eql('Log out');
 });
 
+test('when user is authenticated - should display log in text once log out link is clicked', async (t) => {
+  await pageSetup({ t, withAuth: true, withLoggedInCookies: true });
+
+  const logoutComponent = Selector('[data-test-id="login-logout-component"] a');
+  await t
+    .expect(logoutComponent.exists).eql(true)
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: can we use `.ok()` instead

File: app/pages/homepage/ui.test.js [hunk 4]
```diff
@@ -54,47 +63,98 @@ test('should navigate to home page header banner', async (t) => {
...
+    .click(logoutComponent)
+    .expect(Selector('[data-test-id="login-logout-component"] a').withText('Log in').exists).eql(true)
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: can we just check the inner text of the `logoutComponent` is `Log in`.

File: app/pages/homepage/ui.test.js [hunk 5]
```diff
@@ -54,47 +63,98 @@ test('should navigate to home page header banner', async (t) => {
...
+});
+
+test
+  .requestHooks(logger)('when user is authenticated - should delete cookies once log out link is clicked', async (t) => {
+    await pageSetup({ t, withAuth: true, withLoggedInCookies: true });
+
+    const logoutComponent = Selector('[data-test-id="login-logout-component"] a');
+
+    await t
+      .expect(logoutComponent.exists).eql(true).debug()
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: `.debug()`? also use `.ok()`

File: app/pages/homepage/ui.test.js [hunk 6]
```diff
@@ -54,47 +63,98 @@ test('should navigate to home page header banner', async (t) => {
...
+  });
+
+test
+  .requestHooks(logger)('when user is authenticated - should navigate to / once log out link is clicked', async (t) => {
+    await pageSetup({ t, withAuth: true, withLoggedInCookies: true });
+
+    const logoutComponent = Selector('[data-test-id="login-logout-component"] a');
+
+    await t
+      .expect(logoutComponent.exists).eql(true).debug()
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: .debug()? also use .ok()

File: app/pages/homepage/ui.test.js [hunk 7]
```diff
@@ -54,47 +63,98 @@ test('should navigate to home page header banner', async (t) => {
...
+test('when user is authenticated - should navigate to the identity server log in page when clicking the log in link after logging out', async (t) => {
+  await pageSetup({ t, withAuth: true, withLoggedInCookies: true });
+  nock('http://identity-server')
+    .get('/login')
+    .reply(200);
+
+  const logoutComponent = Selector('[data-test-id="login-logout-component"] a');
+  await t
+    .expect(logoutComponent.exists).eql(true)
+    .click(logoutComponent).debug()
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: `.debug()`?

File: app/pages/homepage/ui.test.js [hunk 8]
```diff
@@ -54,47 +63,98 @@ test('should navigate to home page header banner', async (t) => {
...
+
+test('when user is authenticated - should navigate to the identity server log in page when clicking the log in link after logging out', async (t) => {
+  await pageSetup({ t, withAuth: true, withLoggedInCookies: true });
+  nock('http://identity-server')
+    .get('/login')
+    .reply(200);
+
+  const logoutComponent = Selector('[data-test-id="login-logout-component"] a');
+  await t
+    .expect(logoutComponent.exists).eql(true)
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: `.ok()` please

File: app/pages/homepage/ui.test.js [hunk 9]
```diff
@@ -54,47 +63,98 @@ test('should navigate to home page header banner', async (t) => {
...
+    .click(logoutComponent).debug()
+    .expect(Selector('[data-test-id="login-logout-component"] a').withText('Log in').exists).eql(true)
+    .expect(Selector('[data-test-id="login-logout-component"] a').withText('Log out').exists).eql(false)
```
Comment (alibombaye) [inline app/pages/homepage/ui.test.js]: can we use `extractInnerText` to check it should say `Log in`

File: app/routes.js [hunk 10]
```diff
@@ -35,8 +35,22 @@ export const routes = (authProvider) => {
...
+  });
+
+  router.get('/signout-callback-oidc', async (req, res) => {
+    req.logout();
+    req.session = null;
+
+    if (req.headers.cookie) {
+      req.headers.cookie.split(';')
+        .map(cookie => cookie.split('=')[0])
+        .forEach(cookie => res.clearCookie(cookie));
```
Comment (alibombaye) [inline app/routes.js]: can we call this one `cookieKey` perhaps?

File: app/routes.test.js [hunk 11]
```diff
@@ -73,13 +75,56 @@ describe('routes', () => {
...
   describe('GET /logout', () => {
-    it('should return the text', () => (
-      request(setUpFakeApp())
-        .get('/logout')
+    it('should redirect to the url provided by authProvider', async () => request(setUpFakeApp())
+      .get('/logout')
+      .expect(302)
+      .then((res) => {
+        expect(res.redirect).toEqual(true);
+        expect(res.headers.location).toEqual('/signout-callback-oidc');
```
Comment (alibombaye) [inline app/routes.test.js]: 👍

File: app/test-utils/helper.js [hunk 12]
```diff
@@ -21,5 +21,18 @@ export const getCsrfTokenFromGet = async (app, getPath) => {
...
-  return { cookies, csrfToken };
+
+  return { cookies, csrfToken, app };
+};
+
+export const setFakeCookie = async (app, getPath) => {
+  let cookies;
+  await request(app)
+    .get(getPath)
+    .set('Cookie', ['ddd=eee'])
```
Comment (alibombaye) [inline app/test-utils/helper.js]: can we use something else here rather than `ddd=eee`

Comment (kat1906) [reply app/test-utils/helper.js]: yes (but it makes no difference)

File: app/pages/homepage/ui.test.js (hunk 4 above)
Comment (kat1906) [reply app/pages/homepage/ui.test.js]: This is checking that?

File: app/pages/homepage/ui.test.js (hunk 9 above)
Comment (kat1906) [reply app/pages/homepage/ui.test.js]: this test does that

Comment (alibombaye) [reply app/pages/homepage/ui.test.js]: I know just we so we have a single way to check the text of a component. Unless there is a specific reason for doing it like this.

File: app/test-utils/helper.js (hunk 12 above)
Comment (alibombaye) [reply app/test-utils/helper.js]: I know but call it something like `someFakeCookie` so it reads a little better.

**Pull request merged**